build.webkit.org bots should upload archives directly to S3

briannafan opened the pull request:
#### 62e3e1d1b472669a287ee124839475ad4fe07cbd
<pre>
build.webkit.org bots should upload archives directly to S3
<a href="https://bugs.webkit.org/show_bug.cgi?id=270655">https://bugs.webkit.org/show_bug.cgi?id=270655</a>
<a href="https://rdar.apple.com/124300568">rdar://124300568</a>

Reviewed by Aakash Jain and Jonathan Bedard.

Use GenerateS3URL and UploadFileToS3 in CompileWebKit to upload archives to S3.

* Tools/CISupport/Shared/generate-s3-url:
(main): Adds --minified argument and corresponding bucket.

* Tools/CISupport/build-webkit-org/steps.py:
(ConfigureBuild.__init__):
(ConfigureBuild.start): Adds triggers as a build property so it can be accessed in CompileWebKit.
(DownloadBuiltProduct.run): Allows downloading from master on test instances.
(CompileWebKit.evaluateCommand):
(GenerateS3URL.__init__):
(GenerateS3URL.run):

* Tools/CISupport/build-webkit-org/factories.py:
(Factory.__init__): Adds triggers argument.
(BuildFactory.__init__): Removes archive, upload, and transfer steps. These are added in CompileWebKit.
Also adds triggers argument so that it can be set as a property.

* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):

Canonical link: <a href="https://commits.webkit.org/276500@main">https://commits.webkit.org/276500@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/662d6ccbb7254847c40fa7077fb5f8abdafd29a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47330 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47495 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/40846 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47144 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/27994 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21335 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45419 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/27994 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17898 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/27994 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/39753 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2888 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/27994 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/40045 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49160 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/19809 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/16362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/44882 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/21127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42569 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6211 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/20802 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -459,8 +460,29 @@ def _addToLog(self, logName, message):
+            steps_to_add.extend([UploadBuiltProduct(), TransferToS3(terminate_build=True)])
+
+        # Minified build archive
+        if self.getProperty('platform') in self.APPLE_PLATFORMS or rc in (SUCCESS, WARNINGS) and self.getProperty('user_provided_git_hash'):
```
should it be `or` or `and`

briannafan replied:
Based on what I read in the original `BuildFactory`, we want a minified build archive when we're on an Apple platform. And then combined with the condition in `evaluateCommand`, we get:

`(if platform in apple) OR (rc in success, warnings AND user_provided_git_hash)` which I believe is correct.

aj062 reviewed (approved):
r+ with a comment

aj062 reviewed (changes requested):
Upload steps should continue to be guarded by `if triggers:'.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -459,8 +460,35 @@ def _addToLog(self, logName, message):
+            steps_to_add += [ArchiveBuiltProduct()]
+            if CURRENT_HOSTNAME in BUILD_WEBKIT_HOSTNAMES + TESTING_ENVIRONMENT_HOSTNAMES:
+                steps_to_add.extend([
+                    GenerateS3URL(f"{self.getProperty('fullPlatform')}-{self.getProperty('architecture')}-{self.getProperty('configuration')}"),
```
Maybe let's do all these self.getProperty() calls at one place (where you define triggers). It would make the code more readable.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -459,8 +462,35 @@ def _addToLog(self, logName, message):
+                ])
+            else:
+                # S3 might not be configured on local instances, achieve similar functionality without S3.
+                steps_to_add.extend([UploadBuiltProduct(), TransferToS3(terminate_build=True)])
```
If S3 is not configured, then we shouldn't run TransferToS3() step either. We should remove TransferToS3 from here.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -459,8 +462,35 @@ def _addToLog(self, logName, message):
+                    ])
+                else:
+                    # S3 might not be configured on local instances, achieve similar functionality without S3.
+                    steps_to_add.extend([UploadMinifiedBuiltProduct(), TransferToS3(terminate_build=True)])
```
Ditto. If S3 is not configured, then we shouldn't run TransferToS3() step either. We should remove TransferToS3 from here.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -459,8 +462,35 @@ def _addToLog(self, logName, message):
+
+            # Minified build archive
+            full_platform = self.getProperty('fullPlatform')
+            if (full_platform.startswith('mac') or full_platform.startswith('ios-simulator') or full_platform.startswith('tvos-simulator')
```
Nit: we can make this startswith call compact, as startswith can take a tuple. See: https://docs.python.org/3/library/stdtypes.html#str.startswith "prefix can also be a tuple of prefixes to look for".

e.g.: if (full_platform.startswith(('mac', 'ios-simulator', 'tvos-simulator', 'watchos-simulator'))

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -459,8 +462,35 @@ def _addToLog(self, logName, message):
+            # Minified build archive
+            full_platform = self.getProperty('fullPlatform')
+            if (full_platform.startswith('mac') or full_platform.startswith('ios-simulator') or full_platform.startswith('tvos-simulator')
+               or full_platform.startswith('watchos-simulator')) or (rc in (SUCCESS, WARNINGS) and self.getProperty('user_provided_git_hash')):
```
There is still a small behaviour change when `user_provided_git_hash` is provided. So, the use-case is that someone want to build an archive for a specific hash and manually starts a build, in that case we want to ensure that we upload the archive, irrespective of whether that builder has triggers configured or not.

So, to achieve the similar logic, you can simplify this if condition to just have the `full_platform.startswith`. Then, for the `if triggers` part, add something like this:
```
elif (rc in (SUCCESS, WARNINGS) and self.getProperty('user_provided_git_hash')):
    steps_to_add.extend([
                        GenerateS3URL(f"{self.getProperty('fullPlatform')}-{self.getProperty('architecture')}-{self.getProperty('configuration')}", minified=True),
                        UploadFileToS3(f"WebKitBuild/minified-{self.getProperty('configuration')}.zip", links={self.name: 'Minified Archive'}),
                    ])
```

The reasoning being, that if there are triggers configured, we anyways upload an archive, but if not, than in case of `user_provided_git_hash ` we upload the minified archive (uploading either minified or full archive is fine for `user_provided_git_hash`). @ryanhaddad please let me know if I am missing something for that use-case.

briannafan replied:
Do we want to have a full archive _and_ a minified archive in the case of there being triggers and `user_provided_git_hash`? I think that's how the existing logic works, rather than either minified or full archive being fine. If it doesn't matter whether the user gets full or minified then I'll implement it your way.

briannafan replied:
This also minimizes code duplication (would replace the elif statement you proposed):
```python
if (triggers and full_platform.startswith(('mac', 'ios-simulator', 'tvos-simulator', 'watchos-simulator'))) 
    or (rc in (SUCCESS, WARNINGS) and self.getProperty('user_provided_git_hash')):
            # Minified build archive
            steps_to_add += [ArchiveMinifiedBuiltProduct()]
            if CURRENT_HOSTNAME in BUILD_WEBKIT_HOSTNAMES + TESTING_ENVIRONMENT_HOSTNAMES:
                steps_to_add.extend([
                    GenerateS3URL(f"{full_platform}-{architecture}-{configuration}", minified=True),
                    UploadFileToS3(f"WebKitBuild/minified-{configuration}.zip", links={self.name: 'Minified Archive'}),
                ])
            else:
                steps_to_add.extend([UploadMinifiedBuiltProduct()])
```

aj062 replied:
This looks good.

aj062 reviewed (approved):
Mostly looks good. r+ with comments.

briannafan committed:
build.webkit.org bots should upload archives directly to S3
https://bugs.webkit.org/show_bug.cgi?id=270655
rdar://124300568

Reviewed by Aakash Jain and Jonathan Bedard.

Use GenerateS3URL and UploadFileToS3 in CompileWebKit to upload archives to S3.

* Tools/CISupport/Shared/generate-s3-url:
(main): Adds --minified argument and corresponding bucket.

* Tools/CISupport/build-webkit-org/steps.py:
(ConfigureBuild.__init__):
(ConfigureBuild.start): Adds triggers as a build property so it can be accessed in CompileWebKit.
(DownloadBuiltProduct.run): Allows downloading from master on test instances.
(CompileWebKit.evaluateCommand):
(GenerateS3URL.__init__):
(GenerateS3URL.run):

* Tools/CISupport/build-webkit-org/factories.py:
(Factory.__init__): Adds triggers argument.
(BuildFactory.__init__): Removes archive, upload, and transfer steps. These are added in CompileWebKit.
Also adds triggers argument so that it can be set as a property.

* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):

Canonical link: https://commits.webkit.org/276500@main
